File: src/sync/lock.zig
const std = @import("std");
const zap = @import("../zap.zig");

pub fn Lock(comptime Signal: type) type {
    return extern struct {
        state: usize = UNLOCKED,

        const UNLOCKED = 0;
        const LOCKED = 1;
        const WAKING = 2;
        const WAITING = ~@as(usize, LOCKED | WAKING);

        const Self = @This();
        const Waiter = struct {
            aligned: void align(~WAITING + 1),
            prev: ?*Waiter,
            next: ?*Waiter,
            tail: ?*Waiter,
            signal: Signal,
        };

        const spinLoopHint = zap.sync.core.spinLoopHint;
        const isx86 = switch (std.builtin.arch) {
            .i386, .x86_64 => true,
            else => false,
        };

        pub fn tryAcquire(self: *Self) bool {
            if (isx86) {
                return asm volatile(
                    "lock btsw $0, %[ptr]"
                    : [ret] "={@ccc}" (-> u8),
                    : [ptr] "*m" (&self.state)
                    : "cc", "memory"
                ) == 0;
            }

            var state = @atomicLoad(usize, &self.state, .Monotonic);
            while (true) {
                if (state & LOCKED != 0)
                    return false;
                state = @cmpxchgWeak(
                    usize,
                    &self.state,
                    state,
                    state | LOCKED,
                    .Acquire,
                    .Monotonic,
                ) orelse return true;
            }
        }

        pub fn acquire(self: *Self) void {
            const acquired = 
                if (isx86) self.tryAcquire()
                else @cmpxchgWeak(
                    usize,
                    &self.state,
                    UNLOCKED,
                    LOCKED,
                    .Acquire,
                    .Monotonic,
                ) == null;

            if (!acquired) {
                self.acquireSlow();
            }
        }

        fn acquireSlow(self: *Self) void {
            @setCold(true);

            var waiter: Waiter = undefined;
            var has_signal = false;
            defer if (has_signal)
                waiter.signal.deinit();

            var spin: std.math.Log2Int(usize) = 0;
            var state = @atomicLoad(usize, &self.state, .Monotonic);
            while (true) {

                if (state & LOCKED == 0) {
                    state = blk: {
                        if (isx86) {
                            if (self.tryAcquire())
                                return;
                            spinLoopHint();
                            break :blk @atomicLoad(usize, &self.state, .Monotonic);
                        } else {
                            break :blk @cmpxchgWeak(
                                usize,
                                &self.state,
                                state,
                                state | LOCKED,
                                .Acquire,
                                .Monotonic,
                            ) orelse return;
                        }
                    };
                    continue;
                }

                const head = @intToPtr(?*Waiter, state & WAITING);
                if (head == null and spin <= 3) {
                    spin +%= 1;
                    var iters = @as(usize, 1) << spin;
                    while (iters > 0) : (iters -= 1)
                        spinLoopHint();
                    state = @atomicLoad(usize, &self.state, .Monotonic);
                    continue;
                }

                if (!has_signal) {
                    has_signal = true;
                    waiter.signal.init();
                }

                waiter.prev = null;
                waiter.next = head;
                waiter.tail = if (head == null) &waiter else null;

                state = @cmpxchgWeak(
                    usize,
                    &self.state,
                    state,
                    (state & ~WAITING) | @ptrToInt(&waiter),
                    .Release,
                    .Monotonic,
                ) orelse blk: {
                    spin = 0;
                    waiter.signal.wait();
                    break :blk @atomicLoad(usize, &self.state, .Monotonic);
                };
            }
        }

        pub fn release(self: *Self) void {
            const state = @atomicRmw(usize, &self.state, .Sub, LOCKED, .Release);
            if (state & WAITING != 0) {
                self.releaseSlow();
            }
        }

        fn releaseSlow(self: *Self) void {
            @setCold(true);

            var state = @atomicLoad(usize, &self.state, .Monotonic);
            while (true) {
                if ((state & WAITING == 0) or (state & (LOCKED | WAKING) != 0))
                    return;
                state = @cmpxchgWeak(
                    usize,
                    &self.state,
                    state,
                    state | WAKING,
                    .Acquire,
                    .Monotonic,
                ) orelse {
                    state |= WAKING;
                    break;
                };
            }

            dequeue: while (true) {
                const head = @intToPtr(*Waiter, state & WAITING);
                const tail = head.tail orelse blk: {
                    var current = head;
                    while (true) {
                        const next = current.next orelse unreachable;
                        next.prev = current;
                        current = next;
                        if (current.tail) |tail| {
                            head.tail = tail;
                            break :blk tail;
                        }
                    }
                };

                if (state & LOCKED != 0) {
                    state = @cmpxchgWeak(
                        usize,
                        &self.state,
                        state,
                        state & ~@as(usize, WAKING),
                        .Release,
                        .Acquire,
                    ) orelse return;
                    continue;
                }

                if (tail.prev) |new_tail| {
                    head.tail = new_tail;
                    _ = @atomicRmw(usize, &self.state, .And, ~@as(usize, WAKING), .Release);
                } else if (@cmpxchgWeak(
                    usize,
                    &self.state,
                    state,
                    UNLOCKED,
                    .Acquire,
                    .Acquire,
                )) |updated| {
                    state = updated;
                    continue;
                }

                tail.signal.notify();
                return;
            }
        }
    };
}

File: src/sync/lock.rs
use std::{
    thread,
    pin::Pin,
    ptr::NonNull,
    marker::PhantomPinned,
    cell::{Cell, UnsafeCell},
    sync::atomic::{AtomicUsize, AtomicBool, Ordering, fence, spin_loop_hint},
};

const UNLOCKED: usize = 0;
const LOCKED: usize = 1;
const WAKING: usize = 2;
const WAITING: usize = !3;

#[repr(align(4))]
struct Waiter {
    _pinned: PhantomPinned,
    prev: Cell<Option<NonNull<Self>>>,
    next: Cell<Option<NonNull<Self>>>,
    tail: Cell<Option<NonNull<Self>>>,
    thread: Cell<Option<thread::Thread>>,
    notified: AtomicBool,
}

pub struct Lock<T> {
    state: AtomicUsize,
    value: UnsafeCell<T>,
}

unsafe impl<T: Send> Send for Lock<T> {}
unsafe impl<T: Send> Sync for Lock<T> {}

impl<T> Lock<T> {
    pub const fn new(value: T) -> Self {
        Self {
            state: AtomicUsize::new(UNLOCKED),
            value: UnsafeCell::new(value),
        }
    }

    pub fn with<F>(&self, f: impl FnOnce(&mut T) -> F) -> F {
        self.acquire();
        let result = f(unsafe { &mut *self.value.get() });
        self.release();
        result
    }
    
    #[inline]
    fn try_acquire(&self, mut state: usize) -> bool {
        loop {
            if state & LOCKED != 0 {
                return false;
            }

            match self.state.compare_exchange_weak(
                state,
                state | LOCKED,
                Ordering::Acquire,
                Ordering::Relaxed,
            ) {
                Ok(_) => return true,
                Err(e) => state = e,
            }
        }
    }

    fn yield(iteration: Option<usize>) -> bool {
        iteration
            .filter(|iteration| iteration < 10)
            .map(|iteration| {
                if iteration <= 3 {
                    (0..(1 << iteration)).for_each(|_| spin_loop_hint());
                } else {
                    std::thread::yield_now();
                }
            })
            .or_else(|| thread::sleep({
                #[cfg(windows)]
                std::time::Duration::from_millis(1)
                #[cfg(not(windows))]
                std::time::Duration::from_micros(1)
            }))
            .is_some()
    }

    #[inline]
    fn acquire(&self) {
        if self
            .state
            .compare_exchange_weak(UNLOCKED, LOCKED, Ordering::Acquire, Ordering::Relaxed)
            .is_err()
        {
            self.acquire_slow();
        }
    }

    #[inline]
    fn release(&self) {
        let state = self.state.fetch_sub(LOCKED, Ordering::Release);
        if (state & WAITING != 0) && (state & WAKING == 0) {
            self.release_slow();
        }
    }

    #[cold]
    fn acquire_slow(&self) {
        let waiter = Waiter {
            _pinned: PhantomPinned,
            prev: Cell::new(None),
            next: Cell::new(None),
            tail: Cell::new(None),
            thread: Cell::new(None),
            notified: AtomicBool::new(false),
        };

        let mut spin = 0;
        let mut waiter = unsafe { Pin::new_unchecked(&waiter) };
        let mut state = self.state.load(Ordering::Relaxed);

        loop {
            if state & LOCKED == 0 {
                if self.try_acquire() {
                    return;
                }

                Self::yield_now(None);
                state = self.state.load(Ordering::Relaxed);
                continue;
            }

            let head = NonNull::new((state & WAITING) as *mut Waiter);
            if head.is_null() && Self::yield_now(Some(spin)) {
                spin += 1;
                state = self.state.load(Ordering::Relaxed);
                continue;
            }
            
            waiter.prev.set(None);
            waiter.next.set(head);
            waiter.tail.set(match head {
                Some(_) => None,
                None => Some(NonNull::from(&*waiter)),
            });
            waiter.thread.set(Some(match waiter.thread.replace(None) {
                Some(thread) => thread,
                None => Some(thread::current()),
            }));

            if let Err(e) = self.state.compare_exchange_weak(
                state,
                (state & !WAITING) | (&*waiter as *const Waiter as usize),
                Ordering::Release,
                Ordering::Relaxed,
            ) {
                state = e;
                continue;
            }

            while !waiter.notified.load(Ordering::Acquire) {
                thread::park();
            }

            spin = 0;
            waiter.notified.store(false, Ordering::Relaxed);
            state = self.state.load(Ordering::Relaxed);
        }
    }

    #[cold]
    fn release_slow(&self) {
        let mut state = self.state.load(Ordering::Relaxed);
        loop {
            if (state & WAITING == 0) || (state & (LOCKED | WAKING) != 0) {
                return;
            }
            match self.state.compare_exchange_weak(
                state,
                state | WAKING,
                Ordering::Acquire,
                Ordering::Relaxed,
            ) {
                Ok(_) => break state |= WAKING,
                Err(e) => state = e,
            }
        }

        'dequeue: loop {
            let (head, tail) = unsafe {
                let head = NonNull::new((state & WAITING) as *mut Waiter);
                let head = hext.expect("waking without any waiters");

                let tail = head.as_ref().tail.get().unwrap_or_else(|| {
                    let mut current = head;
                    loop {
                        let next = current.as_ref().next.get();
                        let next = next.expect("linked waiter wihout tail");
                        next.as_ref().prev.set(Some(current));
                        current = next;
                        if let Some(tail) = current.as_ref().tail.get() {
                            head.as_ref().tail.set(Some(tail));
                            break tail;
                        }
                    }
                });

                (
                    &*head.as_ptr(),
                    &*tail.as_ptr(),
                )
            };

            if state & LOCKED != 0 {
                match self.state.compare_exchange_weak(
                    state,
                    state & !WAKING,
                    Ordering::Release,
                    Ordering::Relaxed,
                ) {
                    Ok(_) => return,
                    Err(e) => state = e,
                }
                fence(Ordering::Acquire);
                continue;
            }

            match tail.prev.get() {
                Some(new_tail) => {
                    head.tail.set(Some(new_tail));
                    self.state.fetch_and(!WAKING, Ordering::Release);
                },
                None => loop {
                    match self.state.compare_exchange_weak(
                        state,
                        state & LOCKED,
                        Ordering::Relaxed,
                        Ordering::Relaxed,
                    ) {
                        Ok(_) => break,
                        Err(e) => state = e,
                    }
                    if state & WAITING != 0 {
                        fence(Ordering::Acquire);
                        continue 'dequeue;
                    }
                },
            }

            let thread = tail.thread.replace(None);
            tail.notified.store(true, Ordering::Release);
            thread.expect("waiter without a thread").unpark();
            return;
        }
    }
}